cradle-cassandra/src/main/java/com/exactpro/cradle/cassandra/dao/messages/GroupedMessageIteratorProvider.java

```diff
@@ -94,12 +104,78 @@ private CassandraGroupedMessageFilter createInitialFilter(GroupedMessageFilter f
 				book.getId().getName(),
 				getFirstPage().getId().getName(),
 				filter.getGroupName(),
-				filter.getFrom(),
-				filter.getTo(),
+				leftBoundFilter,
+				rightBoundFilter,
 				order,
 				filter.getLimit());
 	}
 
+	private FilterForGreater<Instant> createLeftBoundFilter(GroupedMessageFilter filter) throws CradleStorageException
+	{
+		FilterForGreater<Instant> result = filter.getFrom();
+		firstPage = FilterUtils.findFirstPage(filter.getPageId(), result, book);
+		Instant leftBoundFromPage = firstPage.getStarted();
+		if (result == null || (filter.getPageId() != null && leftBoundFromPage.isAfter(result.getValue())))
+			return FilterForGreater.forGreaterOrEquals(leftBoundFromPage);
+
+		// If the page wasn't specified in the filter, we should find a batch with a lower date,
+		// which may contain messages that satisfy the original condition
+		LocalDateTime leftBoundLocalDate = TimeUtils.toLocalTimestamp(result.getValue());
+		LocalTime nearestBatchTime = getNearestBatchTime(
+				firstPage,
+				filter.getGroupName(),
+				leftBoundLocalDate.toLocalDate(),
+				leftBoundLocalDate.toLocalTime());
+
+		if (nearestBatchTime != null)
+		{
+			Instant nearestBatchInstant = TimeUtils.toInstant(leftBoundLocalDate.toLocalDate(), nearestBatchTime);
+			if (nearestBatchInstant.isBefore(result.getValue()))
+				result = FilterForGreater.forGreaterOrEquals(nearestBatchInstant);
+			firstPage = FilterUtils.findFirstPage(filter.getPageId(), result, book);
```

Comment:
We have already found the first page in getNearestBatchTime

Reply:
Did we? We have changed the `result` variable so we need to find the page that corresponds to that new value